### PR TITLE
Adding base64 data handling to ping update command

### LINKED ISSUE #1204: Add dataBase64 flag for Ping Source
### Feature request

For parity to be able to create a ping source using the kn cli

### Use case
Use the `dataBase64` for binary data as shown in the [docs](https://knative.dev/docs/eventing/samples/ping-source/)

`kn source ping create test-ping-source --schedule "*/2 * * * *" --dataBase64: "ZGF0YQ==" --sink ksvc:event-display`

### PR DESCRIPTION
**Pull request by vyasgun**
## Description

Adding base64 data handling to ping update command. Handling comments from #1388 

## Changes

* Adding a test case to test builder functions
* Handling base64 data in update command
* Modified usage/error strings for more clarity

## Reference

Fixes #1204

### CODE REVIEW

File: pkg/kn/commands/source/ping/create.go [hunk 1]
```diff
@@ -76,21 +70,8 @@ func NewPingCreateCommand(p *commands.KnParams) *cobra.Command {
...
-				}
-			case base64Encoding:
-				dataBase64 = updateFlags.data
-			case textEncoding:
-				data = updateFlags.data
-			default:
-				err := fmt.Errorf("invalid value: %v", updateFlags.encoding)
+			data, dataBase64, err := getDataFields(&updateFlags)
+			if err != nil {
 				fmt.Fprintf(cmd.OutOrStderr(), "Ping source creation failed with error: %v", err)
```
Comment (dsimansk) [inline pkg/kn/commands/source/ping/create.go:75]: The message format should be probably aligned with the rest of sources.
E.g. see: https://github.com/knative/client/blob/main/pkg/kn/commands/source/container/create.go#L64-L68

**Comment by dsimansk** [comment]
I'd be great if you could add a few test cases to E2E tests as well. At least asserting no errors on create and binary data in description. Bonus would be asserting sink logs. :)
https://github.com/knative/client/blob/main/test/e2e/ping_test.go#L29

Finally, please add new feature entry to `CHANGELOG.adoc` file.

File: CHANGELOG.adoc [hunk 2]
```diff
@@ -17,6 +17,12 @@
 |===
 | | Description | PR
 
+
+| ✨
+| Adding base64 data handling to ping update command
+| https://github.com/knative/client/pull/1392[#1392]
+| https://github.com/knative/client/pull/1388[#1388]
```
Comment (dsimansk) [inline CHANGELOG.adoc:24]: This should be rather one line, otherwise it breaks the table rendering.

E.g.:
```suggestion
| https://github.com/knative/client/pull/1388[#1388], https://github.com/knative/client/pull/1392[#1392]
```

File: pkg/kn/commands/source/ping/describe.go [hunk 3]
```diff
@@ -117,7 +117,11 @@ func NewPingDescribeCommand(p *commands.KnParams) *cobra.Command {
 func writePingSource(dw printers.PrefixWriter, source *clientsourcesv1beta2.PingSource, printDetails bool) {
 	commands.WriteMetadata(dw, &source.ObjectMeta, printDetails)
 	dw.WriteAttribute("Schedule", source.Spec.Schedule)
-	dw.WriteAttribute("Data", source.Spec.Data)
+	if source.Spec.DataBase64 != "" {
```
Comment (dsimansk) [inline pkg/kn/commands/source/ping/describe.go:120]: A simple test case would be nice as well, pretty please :pleading_face:.

File: test/e2e/ping_test.go [hunk 4]
```diff
@@ -64,6 +64,19 @@ func TestSourcePing(t *testing.T) {
 	mymsg := "This is a message from Ping."
 	pingSourceCreate(r, "testpingsource3", "*/1 * * * *", mymsg, "ksvc:testsvc1")
 	verifyPingSourceDescribe(r, "testpingsource3", "*/1 * * * *", mymsg, "testsvc1")
+
+	t.Log("create Ping source with base64 data without encoding flag")
+
```
Comment (dsimansk) [inline test/e2e/ping_test.go:69]:
```suggestion
```

File: test/e2e/ping_test.go [hunk 5]
```diff
@@ -64,6 +64,19 @@ func TestSourcePing(t *testing.T) {
...
+	pingSourceCreate(r, "testpingsource4", "* * * * */1", "cGluZw==", "ksvc:testsvc1")
+	verifyPingSourceDescribe(r, "testpingsource4", "* * * * */1", "cGluZw==", "testsvc1")
+
+	t.Log("create Ping source with text data and encoding flag")
+	pingSourceCreateWithEncoding(r, "testpingsource6", "* * * * */1", "cGluZw==", "ksvc:testsvc1", "text")
+	verifyPingSourceDescribe(r, "testpingsource6", "* * * * */1", "cGluZw==", "testsvc1")
```
Comment (dsimansk) [inline test/e2e/ping_test.go:75]: Is it intentional to add base64 data and "--encoding text", because description says "text data"?

File: test/e2e/ping_test.go [hunk 6]
```diff
@@ -64,6 +64,19 @@ func TestSourcePing(t *testing.T) {
...
+
+	t.Log("update Ping source with base64 data and encoding flag")
+	pingSourceUpdateData(r, "testpingsource6", "cGluZw==", "base64")
+	verifyPingSourceDescribe(r, "testpingsource6", "* * * * */1", "cGluZw==", "testsvc1")
```
Comment (dsimansk) [inline test/e2e/ping_test.go:79]: Can you please add additional case for invalid `--encoding` value.

File: test/e2e/ping_test.go (hunk 5 above)
Comment (vyasgun) [reply test/e2e/ping_test.go:75]: No, I could add any data. I have added a unit test that shows explicit encoding specification overrides the automatic interpretation the `--data` flag does and added a similar test case.

File: pkg/kn/commands/source/ping/describe.go (hunk 3 above)
Comment (vyasgun) [reply pkg/kn/commands/source/ping/describe.go:120]: Added one 😆

File: test/e2e/ping_test.go (hunk 5 above)
Comment (dsimansk) [reply test/e2e/ping_test.go:74]: OK, makes senses now. :)

**Comment by dsimansk** [comment]
/lgtm

Not approving yet for others to take a look maybe.

File: docs/cmd/kn_source_ping_create.md [hunk 7]
```diff
@@ -18,8 +18,8 @@ kn source ping create NAME --sink SINK
 
 ```
       --ce-override stringArray   Cloud Event overrides to apply before sending event to sink. Example: '--ce-override key=value' You may be provide this flag multiple times. To unset, append "-" to the key (e.g. --ce-override key-).
-  -d, --data string               Json data to send
-  -e, --encoding string           Preferred encoding format. Valid values: text/base64
+  -d, --data string               Json data to send. Can determine if data is plain text or base64 encoded
```
Comment (dsimansk) [inline docs/cmd/kn_source_ping_create.md:21]: `Json data to send. Can determine if data is plain text or base64 encoded`

Maybe I'd rephrase the whole to read a bit better, but not sure how exactly. I'll try to put some thoughts to it. :)

Comment (vyasgun) [reply docs/cmd/kn_source_ping_create.md:21]: I could think of this:

`Json data to send.This flag can implicitly determine the encoding of the supplied data (text | base64)`

**Comment by dsimansk** [comment]
Thanks @vyasgun!

So let's go with approve this time, and leave lgtm for others to take a look. :)
/approve

**Comment by dsimansk** [comment]
@vyasgun thanks again, great job on both related PRs. :+1: 

/lgtm
